### ISSUE TITLE
Cannot conditionally register different instances of the same type for the same service

### ISSUE THREAD

**Issue by Bouke** [simpleinjector/SimpleInjector#836]
I have two instances of the same type, which I want to discriminate between based on the constructor's argument of the consumer. However when registering these conditionals, Simple Injector doesn't accept this and suggests I combine the registrations. This is not possible, as having two different registrations was the whole reason for splitting them up. This looks similar to #485, but in this case the registrations are not the same.

## Expected behaviour
As the registrations are different and the conditionals are mutual exclusive, this should be accepted.

## Actual behaviour
> System.InvalidOperationException: 'There is already a conditional registration for IFoo (with implementation Foo) that overlaps with the conditional registration for Foo that you are trying to make. This new registration causes ambiguity, because both registrations would be used for the same closed service types. You can merge both registrations into a single conditional registration and combine both predicates into one single predicate.'

## To Reproduce

``` c#
interface IFoo { }
class Foo : IFoo { public bool Debug { get; set; } }

container.RegisterConditional<IFoo>(
    Lifestyle.Singleton.CreateRegistration(typeof(IFoo), new Foo { }, container),
    c => c.Consumer.Target.Name.StartsWith("debug", StringComparison.Ordinal));

container.RegisterConditional<IFoo>(
    Lifestyle.Singleton.CreateRegistration(typeof(IFoo), new Foo { Debug = true }, container),
    c => !c.Handled);
```

## Additional context
- Simple Injector 5.0.2

**Comment by dotnetjunkie** [comment]
What you're seeing is IMO a bug.

For now, you can work around this problem by using the `CreateRegistration` overload that accepts a `Func<object>`:

``` c#
container.RegisterConditional<IFoo>(
    Lifestyle.Singleton.CreateRegistration(typeof(IFoo), () => new Foo { }, container),
    c => c.Consumer.Target.Name.StartsWith("debug", StringComparison.Ordinal));

container.RegisterConditional<IFoo>(
    Lifestyle.Singleton.CreateRegistration(typeof(IFoo), () => new Foo { Debug = true }, container),
    c => !c.Handled);
```

**Comment by dotnetjunkie** [comment]
[bug-836](https://github.com/simpleinjector/SimpleInjector/tree/bug-836) branch created.

**Comment by dotnetjunkie** [comment]
Version 5 introduced the new `SingletonLifestyle.CreateRegistration(Type, object, Container)` overload to allow simplifying and optimizing the creating a `Registration` for a "single instance." Unfortunately, the use case of using that within `RegisterConditional` was overlooked, causing this bug to appear.

**Comment by kwlin** [comment]
Thnx @Bouke for the reproduction steps. I'll give it a try to fix this issue 👍

**Comment by dotnetjunkie** [comment]
This has just been fixed in [v5.0.4](https://github.com/simpleinjector/SimpleInjector/releases/tag/5.0.4). Thanks to @kwlin
